diff -bBruPN knife-windows-0.8.2/lib/chef/knife/core/windows_bootstrap_context.rb /root/knife-windows-0.8.2-patched/lib/chef/knife/core/windows_bootstrap_context.rb
--- knife-windows-0.8.2/lib/chef/knife/core/windows_bootstrap_context.rb	2015-01-27 15:50:51.147154392 +0000
+++ /root/knife-windows-0.8.2-patched/lib/chef/knife/core/windows_bootstrap_context.rb	2015-01-27 15:13:07.619991818 +0000
@@ -76,6 +76,40 @@
             client_rb << %Q{encrypted_data_bag_secret "c:/chef/encrypted_data_bag_secret"\n}
           end
 
+          # We configure :verify_api_cert only when it's overridden on the CLI
+          # or when specified in the knife config.
+          if !@config[:node_verify_api_cert].nil? || knife_config.has_key?(:verify_api_cert)
+            value = @config[:node_verify_api_cert].nil? ? knife_config[:verify_api_cert] : @config[:node_verify_api_cert]
+            client_rb << %Q{verify_api_cert #{value}\n}
+          end
+
+          # We configure :ssl_verify_mode only when it's overridden on the CLI
+          # or when specified in the knife config.
+          if @config[:node_ssl_verify_mode] || knife_config.has_key?(:ssl_verify_mode)
+            value = case @config[:node_ssl_verify_mode]
+            when "peer"
+              :verify_peer
+            when "none"
+              :verify_none
+            when nil
+              knife_config[:ssl_verify_mode]
+            else
+              nil
+            end
+
+            if value
+              client_rb << %Q{ssl_verify_mode :#{value}\n}
+            end
+          end
+
+          if @config[:ssl_verify_mode]
+            client_rb << %Q{ssl_verify_mode :#{knife_config[:ssl_verify_mode]}\n}
+          end
+
+          unless trusted_certs.empty?
+            client_rb << %Q{trusted_certs_dir "c:/chef/trusted_certs"\n}
+          end
+
           escape_and_echo(client_rb)
         end
 
